### DISCUSSION TITLE
[Support] 接口返回的若是204状态码，HttpClient访问时会报错 Unable to parse response data。这里说的是@midwayjs/core里的HttpClient

### DISCUSSION THREAD

**Discussion by googoo625** [midwayjs/midway#3983] [Q&A] [answered]
### Describe the problem（描述问题）

接口返回的若是204状态码，HttpClient访问时会报错 ：
[httpclient] Unable to parse response data

这里说的是@midwayjs/core里的HttpClient

状态码204 表示成功完成并且不需要有返回内容，是一个标准Http状态码。

是否可以支持一下。

### Midway Versions（Midway 版本）

"@midwayjs/core": "^3.15.0",

**Comment by czy88840616** [comment]
```
  const result = await httpclient.request(`http://127.1:3000`, {
    method: 'GET',
  });
  expect(result.statusCode).toEqual(204);
```
  不传 dataType 就行了

**Reply by googoo625** [reply]
多次验证也没办法成功，走到 httpclient.request 就报错了。
请问你用的是 @midwayjs/core 中的 HttpClient 吗？

我的版本是：
~~~
"@midwayjs/core": "3.16.2"
~~~

以下是报错时的代码：
~~~
import { HttpClient } from '@midwayjs/core';

const httpclient = new HttpClient({
	timeout: 5000
});
await httpclient.request('http://localhost:7001/api/v1/groups/nmsosvn47034p5v088zljk1v54', {
  method: 'PUT',
  data: {
    name: 'name',
    description: 'description',
  },
  headers: {
    Authorization: 'Bearer xxxxxxxxxx',
  },
  contentType: 'json'
});
~~~

没传dataType。
也测试了GET接口，同样会报错。

**Reply by czy88840616** [reply]
现在手边没有电脑，试试axios？

**Reply by googoo625** [reply]
我就是想问问 midway/core 的 HttpClient 能否改进支持

**Reply by czy88840616** [reply]
我需要验证才能改进，不过最近在外边玩没有拿电脑，等我回国了再看：）

**Reply by waitingsong** [reply]
> 多次验证也没办法成功，走到 httpclient.request 就报错了。 请问你用的是 @midwayjs/core 中的 HttpClient 吗？
> 
> 我的版本是：
> 
> ```
> "@midwayjs/core": "3.16.2"
> ```
> 
> 以下是报错时的代码：
> 
> ```
> import { HttpClient } from '@midwayjs/core';
> 
> const httpclient = new HttpClient({
> 	timeout: 5000
> });
> await httpclient.request('http://localhost:7001/api/v1/groups/nmsosvn47034p5v088zljk1v54', {
>   method: 'PUT',
>   data: {
>     name: 'name',
>     description: 'description',
>   },
>   headers: {
>     Authorization: 'Bearer xxxxxxxxxx',
>   },
>   contentType: 'json'
> });
> ```
> 
> 没传dataType。 也测试了GET接口，同样会报错。

我测试是正常的

![2024-08-16_140312](https://github.com/user-attachments/assets/37608abe-2c93-49b3-8817-72633c04aab6)

**Comment by waitingsong** [comment]
![2024-08-16_140719](https://github.com/user-attachments/assets/7323ac21-6427-4843-a82c-e54c70c24c3a)

**Comment by googoo625** [comment]
唯一区别就是接口不是声明为"@HttpCode(204)"的，而是在中间件Middleware里整体判断并设置的，ctx.status=204。
```
            const result = await next();
            ctx.status = result?.status || 200;
            ctx.logger.info(`[---Api---] ${ctx.method} ${ctx.path} ${Date.now() - ctx.startTime}ms`);
//            return (ctx.status == 204) ? undefined : result;
            return result;
```
只有204会在访问时报错"[httpclient] Unable to parse response data"，其它像 200/400/401/403/404等情况都可以正常访问并得到返回状态码。
不明白怎么回事。

**Comment by waitingsong** [comment]
这样呢？
```js
return ctx.status === 204 ? '' : result;
```

**Reply by googoo625** [reply]
试了，访问时的报错没变：[httpclient] Unable to parse response data

**Reply by waitingsong** [reply]
> 试了，访问时的报错没变：[httpclient] Unable to parse response data

那试试看把 '' 换成其它数据，比如 null, {} 什么的，如果还不行看看相应 content-type 是不是不对

**Reply by czy88840616** [reply]
直接从脚手架创个case让我们看看？

**Reply by googoo625** [reply]
现在我**换成@midwayjs/axios，一切都正常了**。

所以还是HttpClient哪里有点问题？就204时会报错，其他状态码都OK。

准备了一个demo程序，结果运行后一切正常，不能复现.....同样用了Middleware并在中间件里设置ctx.status=204。

而实际环境中，即使在中间件里拦截接口直接返回结果（如下图红框），204时也会报错；返回undefined/null/{}之类也不行。
打印了ctx对象，没发现和demo程序有不同之处。
实际环境的ctx会多加两个变量（user, roleName），似乎不是他们的影响。
实际环境有很多注入service，会影响到接口返回吗？

准备就用axios了，也希望HttpClient能查到哪里的问题。

![17254177235800](https://github.com/user-attachments/assets/40bf008f-0ef9-49b8-826d-20e560c7c579)

**Comment by waitingsong** [comment]
是不是某个中间件修改了响应 header/body

**Reply by googoo625** [reply]
没有，每个服务只有一个中间件，就是上面图中的。而且换成axios就正常了。
HttpClient 处理204时啥因素会造成 Unable to parse response data 错？
更新到 v3.17.1 最新版了，没变化。

**Reply by czy88840616** [reply]
<img width="1109" alt="image" src="https://github.com/user-attachments/assets/06a215e0-aa92-4032-898d-37ebfaa2cb06">
只跟你传递的 dataType 有关。

**Comment by googoo625** [comment]
事情解决了！
访问接口时明确设置dataType="octet"或”text"，不要是"json"，也别不赋值。
之前是没赋值，不行；看了源码试着这么改，就生效了！

仍奇怪的是demo程序中dataType不赋值是正常的，前面waitingsong的测试用例也正常。

```
options.dataType = options.dataType || (!["PUT", "DELETE", "PATCH"].includes(options.method) ? "json" : "octet");
```

**Reply by waitingsong** [reply]
是不是请求的 request header 在某处设定了 data type 值

**Reply by googoo625** [reply]
明白了，是因为重复使用httpClient对象导致的。
HttpClient定义的request方法中，用了Object.assign，这样会改变了defaultOptions的内容。
上一次调用request时若传了dataType=json，则会保留在defaultOptions里，再次调用request时若没传入dataType就会是json了。

所以每次调用都要重新new一个HttpClient对象。
或者改改下面的源码，使Object.assign之前defaultOptions里没有多余的属性。

```
export class HttpClient {
  constructor(
    readonly defaultOptions: Pick<
      HttpClientOptions,
      'headers' | 'timeout' | 'method'
    > = {}
  ) {}

  async request(
    url: string,
    options?: HttpClientOptions
  ): Promise<HttpClientResponse> {
    return makeHttpRequest(url, Object.assign(this.defaultOptions, options));
  }
}

```

**Comment by googoo625** [comment, marked as answer]
问题解决办法：
1）访问204接口时，dataType不要传json
2）避免重复使用HttpClient对象，每次访问都应新建一个。
感谢帮助！

还有，期待HttpClient稍做完善，避免使用Object.assign导致的defaultOptions属性增加，毕竟重复使用httpClient对象的情况是有可能的。

**Reply by czy88840616** [reply]
#4058
